### PR TITLE
Implement cursor blinking

### LINKED ISSUE #302: Cursor Not Blinking
Is there an option to cause the cursor to blink? I find it to be helpful confirmation that the terminal has (or doesn't have) focus.

### PR DESCRIPTION
**Pull request by Dettorer**
> Please make sure to document all user-facing changes in the
> `CHANGELOG.md` file.

Done

> If support for a new escape sequence was added, it should be documented
> in `./docs/escape_support.md`.

Done
note: `CSI ? 12 h` and `CSI ? 12 l` were already marked as supported.

The default blinking rate is 500ms hide -> 500ms show -> …

TODO:
- [x] wether to blink or not by default
- [x] start blinking at startup if configuration says so
- [x] start or stop blinking at vi/normal mode change according to configuration
- [x] handle `CSI 1/3/5 SP q` (set cursor style to blinking block/underline/bar)
- [x] make the "blinking cursor" an option on each cursor style instead of separate new styles
- [x] make the vi cursor blink too

Optionally:
- [x] the blinking rate
- [x] completely disable blinking (ignore escape sequences), possibly by setting the rate to `0`
- [x] reinitialize the blinking cycle on user input
- (?) allow different blinking rate for the default cursor and vi cursor

Fixes #302

### CODE REVIEW

**Comment by chrisduerr** [comment]
Hey @Dettorer, do you still plan to work on this? If you need some help please let me know.

**Comment by Dettorer** [comment]
Yep, vacation got in the way but I still plan to work on this.

File: alacritty.yml
Comment (stertingen) [inline alacritty.yml:330]: I'd prefer to have a separate option for this, like `blink: true`. Here, we mix two different things in a single configuration option.

Maybe in the future, support for custom cursors may be added and someone is crazy enough to want a blinking emoji. That would be easier if I could set something like: `style: '★'` and set the blink option elsewhere.

File: alacritty.yml
Comment (stertingen) [inline alacritty.yml:346]: Or, instead of adding a separate option, set `blink_rate` to `0` to disable blinking.

One more thing: The unit of the value is ms, right? I'd suggest to rename the config option to `blink_period` or make this a proper rate where the default value is 2 Hz. Or, at least, mention somewhere in the comment that the value has the unit ms.

File: alacritty.yml
Comment (Dettorer) [reply alacritty.yml:330]: That would work too yes, I went for this kind of settings because of the escape sequences (not handled yet in the current state of that PR) that instruct the terminal to switch to "Block" or "BlinkingBlock", but that's not a blocker.

File: alacritty.yml
Comment (Dettorer) [reply alacritty.yml:346]: Yes, setting `blink_rate` to 0 does disable blinking, I just forgot to say it in the comment, thanks for the reminder.

The fact that it's in ms is said on the first line of that comment. I personally find a value of ms more common than a frequency in Hz, I rarely see tools that have configuration options in Hz.

Comment (stertingen) [reply alacritty.yml:346]: Oh yeah, my bad. I was too focused on the paragraph, nevermind then.

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:330]: As mentioned previously by @kchibisov, you can actually combine this with the blink rate option.

Comment (chrisduerr) [reply alacritty.yml:330]: > Maybe in the future, support for custom cursors may be added and someone is crazy enough to want a blinking emoji. That would be easier if I could set something like: style: '★' and set the blink option elsewhere.

That's certainly not desired, but generally you are right. This should just be a separate boolean flag.

Comment (Dettorer) [reply alacritty.yml:330]: I can make the `CursorStyle` enum a two-fields struct: `shape` (current enum) and `blinking` (bool).
Or keep the current enum with `Blinking*` variants until alacritty supports more shapes, what would you think is more reasonable @chrisduerr?

Comment (chrisduerr) [reply alacritty.yml:330]: You should probably have a `CursorShape` enum and a `CursorStyle` struct that has a boolean option if the cursor should be blinking or not.

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:325]:
```suggestion
  #style:
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:324]:
```suggestion
  # Cursor style
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:326]:
```suggestion
      # The shape of the cursor
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:334]:
```suggestion
      # Whether the cursor should blink by default or not.
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:345]:
```suggestion
  # Cursor blinking interval
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:352]:
```suggestion
  # Rate in milliseconds used for cursor blinking when a blinking cursor style is active. A `blink_rate` of `0` will completely disable cursor blinking.
```

File: alacritty/src/event.rs
Comment (chrisduerr) [inline alacritty/src/event.rs:759]: The send_event here doesn't need any error handling, just how it's done for other channels in Alacritty.

It also seems pointless to send an event to the event loop, when we are literally right in the event handler, that just seems like a pointless indirection. Can't we just start the blinking ourselves?

File: alacritty/src/event.rs
Comment (chrisduerr) [inline alacritty/src/event.rs:939]: Why define a shorthand for scheduler, but then not use it?

File: alacritty/src/event.rs
Comment (chrisduerr) [inline alacritty/src/event.rs:938]: What if the configuration file is changed to disable blinking while Alacritty is active? We should support live config reload for this.

File: alacritty/src/event.rs
Comment (chrisduerr) [inline alacritty/src/event.rs:950]:
```suggestion
                        processor.ctx.terminal.cursor_blinking_out ^= true;
```

This comment is pointless.

File: alacritty/src/input.rs
Comment (chrisduerr) [inline alacritty/src/input.rs:871]:
```suggestion
        // Disable cursor blinking while the user is typing.
```

File: alacritty/src/input.rs
Comment (chrisduerr) [inline alacritty/src/input.rs:875]: Should this really always be done when any key is pressed, or only when a character is received? I'd argue that just pressing Control shouldn't disable cursor blinking.

File: alacritty/src/input.rs
Comment (chrisduerr) [inline alacritty/src/input.rs:873]: Where is this 500 coming from?

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:1037]: Isn't this a bit late? We could save ourselves a lot of work if we just filtered it out much earlier.

File: alacritty_terminal/src/ansi.rs
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:347]: This comment doesn't really provide any value. `Describes the style of cursor` is the same as `CursorStyle` and the comment of `shape + blinking` is just all the fields. We shouldn't document fields because those might change, so that's actually actively harmful. I think it would be better to just delete this rather than using it as-is.

File: alacritty_terminal/src/ansi.rs
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1098]: I feel like the additional conditional is just unnecessarily complicated. Expanding this to struct to have separate entries for blinking/non-blinking probably makes it a ton easier to just read without having to think about what is going on.

File: alacritty_terminal/src/config/mod.rs
Comment (chrisduerr) [inline alacritty_terminal/src/config/mod.rs:128]: We can't just change things, we need to still support the old format and print deprecation warnings.

File: alacritty_terminal/src/event.rs
Comment (chrisduerr) [inline alacritty_terminal/src/event.rs:18]: I don't think this is a good name, since `Blink` has no clear indication of what it should do and thus would make it difficult to implement. It should instead be something like `ToggleCursorVisibility`.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:62]: This variable name is incredibly confusing, could we go with something more descriptive. A cursor being "out" is not something I would understand, so other people would probably have even more of a problem.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:693]: This variable name should be changed.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:847]: I'm questioning the API ergonomics of how this is supposed to work. The terminal instructs the UI when blinking should start or end, then the UI tells the terminal to tell the UI to update it about how things should work? That seems insanely convoluted.

Why can't the UI just figure this out itself?

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:1327]: The unfocused hollow block cursor never blinks, but it seems like other cursors still blink even without focus.

I think if Alacritty isn't focused, the cursor shouldn't blink. I wouldn't even make this configurable, since it's just a waste of resources really.

Comment (Dettorer) [reply alacritty_terminal/src/term/mod.rs:1327]: Indeed, disabling blinking while unfocused also is the behavior of XTerm and rxvt, I'll try to make it work too.

File: alacritty_terminal/src/term/mod.rs
Comment (Dettorer) [reply alacritty_terminal/src/term/mod.rs:847]: I didn't find a better way to do that, but that's my first contribution so if you think that's convoluted there must be an other way. I'll look into it and ask question on IRC if needed.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/mod.rs:62]: It might be a bit of an indication that the way this is done isn't ideal.

I don't quite remember all our previous conversations, but have you tried storing the cursor visibility for blinking outside of alacritty_terminal? Currently I don't see why it's needed inside of it.

File: alacritty/src/event.rs
Comment (chrisduerr) [reply alacritty/src/event.rs:759]: That sounds like a good idea anyways, since we want to keep our event handler functionally simple and delegate to methods for anything even slightly complex.

File: alacritty_terminal/src/ansi.rs
Comment (chrisduerr) [reply alacritty_terminal/src/ansi.rs]: We can keep it at `Describes the style of the cursor.` I don't really care either way. But I wouldn't remove existing documentation for this.

File: alacritty_terminal/src/ansi.rs
Comment (chrisduerr) [reply alacritty_terminal/src/ansi.rs:1097]: Instead of `1 | 2 => ps == 1`, you'd just do `1 => true, 2 => false`. Highly simplified, but hopefully you get the gist.

File: alacritty_terminal/src/event.rs
Comment (chrisduerr) [reply alacritty_terminal/src/event.rs:18]: Sure, that sounds like a good idea.

**Comment by yuri1969** [comment]
I got feedback using 1adcb53 - when you launch an interactive tool (such as `vi` or `ranger`) the cursor stops blinking and remains so after you return form the app.

File: alacritty_terminal/src/ansi.rs
Comment (nixpulvis) [reply alacritty_terminal/src/ansi.rs:1097]: Something else is wrong around this, since I get the error: "[ERROR] ... Problem with config: missing field \`shape\`; using default" when I insert the following configuration I would expect to work:

```yaml
cursor:
  style:
    blinking: true
```

**Comment by chrisduerr** [comment]
@Dettorer Please let me know when you need another review, that way I don't have to check every time if all the comments have been addressed. Just to make sure you're not waiting for a review while I'm waiting for you to change things.

**Comment by Dettorer** [comment]
Yeah, sorry, I'm planning on marking the opened conversations as resolved as I go fixing or implementing them. I'll ask again explicitly for a review once it's done.

Although I should mention that I might not be able to work on this for a month or two (my job is about to get very time consuming). If someone wants/needs to take over I'm totally ok with it, but I'll also be happy to finish this PR once things calm down on my side.

**Comment by jasonjckn** [comment]
looks like the work here is nearly done? move it to the finish line?

**Comment by Dettorer** [comment]
Feel free to do so if you want to, but I should be able to start working on this again pretty soon.

**Comment by PouriDev** [comment]
Any progress on this PR?

**Comment by Dettorer** [comment]
Ok sorry, the sanitary situation made my job way more time consuming than I anticipated and I keep forgetting to give updates here. I should stop saying "I should be available soon", it's becoming a pattern.

I suggest you consider me out of this for the foreseeable future. I'll still be happy to continue my work on this when I have the time if nobody had a go at it, but I can't say when that'll be.

Again, I should be able to answer some questions from time to time though if someone wants to understand my drafts.

**Comment by kchibisov** [comment]
I think @chrisduerr was willing to pick up this after we do 0.6.0 release, if not I can pick up it.

**Comment by chrisduerr** [comment]
> I think @chrisduerr was willing to pick up this after we do 0.6.0 release, if not I can pick up it.

Will be absolutely trivial to do at the beginning of 0.7.0, yes. Especially since we already have cursor hiding logic in place now.

File: alacritty_terminal/src/ansi.rs
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:878]: We should probably keep support for this I'd assume, even though 1337 should be used instead? I'm not sure which application uses this escape.

Comment (kchibisov) [reply alacritty_terminal/src/ansi.rs]: 1337 sounds like iterm

**Comment by chrisduerr** [comment]
@kchibisov I think this should be ready for review.

**Comment by chrisduerr** [comment]
One thing I've noticed: It seems like a `reset` after blinking has been disabled with an escape sequence doesn't restore the blinking. That still needs to be investigated.

File: alacritty/src/event.rs
Comment (kchibisov) [inline alacritty/src/event.rs:1090]: You unconditionally enable blinking even if the blinking cursor mode was unset before, on focus change, this is noticeable in e.g. weechat which disables blinking completely, and after refocus the cursor starts blinking, which shouldn't be. You should track whether the current mode/cursor allows the cursor blinking in a first place, to handle unset of the mode properly.

File: alacritty/src/input.rs
Comment (kchibisov) [inline alacritty/src/input.rs:906]: Need something like that for Vi motions, since right now it's not really usable in it, since cursor constantly blinking while you're moving and you don't see the cursor.

**Review by kchibisov — CHANGES_REQUESTED** [review]
I've mentioned on IRC, that I think it makes sense to provide an option to disable blinking or disable unsetting blinking cursor(all ncurses apps and some applications don't provide an option to have blinking cursor in them). However I'm not sure how many terminals allow those things, so requires checking some of them.

File: alacritty_terminal/src/config/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/config/mod.rs:164]: I personally find neither 750 nor 500 ideal. So I picked the next best thing in between. Though I might have just been biased based on the other terminals I tested.

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:351]: It's already documented in the comment above vi mode?

File: alacritty_terminal/src/config/mod.rs
Comment (kchibisov) [reply alacritty_terminal/src/config/mod.rs:164]: I think it's nice, since I don't even like the default you've picked, and I'd image that it'll be hard to pick a reasonable one, since you like 666, but I don't find it good. Also, even gtk allows changing it.

Comment (chrisduerr) [reply alacritty_terminal/src/config/mod.rs:164]: Fair enough. Which would be your preferred blink rate? I don't really like blinking to begin with.

File: alacritty/src/event.rs
Comment (chrisduerr) [inline alacritty/src/event.rs:510]: I'm not sure if I like this API. The `cursor_style.blinking(current_state)` thing is a bit weird, but that's the easiest API for us in Alacritty.

I think the other alternative would be a bit nicer, which would be `cursor_style.blinking_override().unwrap_or(terminal_state)`, but it's not as concise for our usage.

What do you think @kchibisov?

**Comment by chrisduerr** [comment]
@kchibisov I think all your suggestions should be integrated now. The only question left is if we want to do anything about the `blink_interval` I suppose. It should be trivial to make that cursor-specific, but I'm not sure if there's any benefit to that?

File: alacritty.yml
Comment (kchibisov) [inline alacritty.yml:362]: I think 700 or 650 should be a bit better for default, since they are no that annoying like 500.

File: alacritty/src/event.rs
Comment (kchibisov) [inline alacritty/src/event.rs:535]: I'm not sure if I like `mode` here, maybe we should use something like `fn started_typing(&mut self)`. Since each time I see input mode I think it's IME stuff.

File: docs/escape_support.md
Comment (kchibisov) [inline docs/escape_support.md:98]: Seems like alignment wasn't reverted after you've removed OSC 1337.

File: alacritty/src/event.rs
Comment (kchibisov) [inline alacritty/src/event.rs:523]: If you have blinking `On` and then you do `printf "\e[?12l\a"`, it's still blinking, but it shouldn't. However it seems like vte literally does nothing on Blinking cursor mode, so maybe it's fine? Like it neither disables it nor enables it, and blinking can only be changed via `CSI Ps SP q`, maybe we should follow it as well? Have you checked how it works in xterm/urxvt?

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/mod.rs:2112]: Separate PR, has nothing to do with this.

File: alacritty/src/event.rs
Comment (chrisduerr) [reply alacritty/src/event.rs:523]: Just a bug.

File: alacritty_terminal/src/event.rs
Comment (chrisduerr) [reply alacritty_terminal/src/event.rs:15]: I think `CursorBlinkingChange` should be a bit better.

File: alacritty/src/event.rs
Comment (chrisduerr) [reply alacritty/src/event.rs:535]: I feel like a `started_typing` method should return a `bool` or something? I don't like that name since it doesn't really help me understand that it does.

File: alacritty_terminal/src/config/mod.rs
Comment (kchibisov) [inline alacritty_terminal/src/config/mod.rs:128]: Do we really want to keep `u64` here? I don't see any reason to use it over something like `u32` or `u16` for blinking.

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: alacritty/src/event.rs
Comment (chrisduerr) [reply alacritty/src/event.rs:1094]: Yes, still a leftover from the previous commits.

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:368]: You mean force clamp it to be at least 50ms? I'm not sure if that actually provides any real benefit?

It's not like your computer suddenly explodes when you put 0 as the interval. Is this realistically something that can be messed up by accident?

File: alacritty_terminal/src/config/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/config/mod.rs:128]: There is a reason: The duration conversion takes a u64. So one way or another we need to convert it to u64 anyways at the only place we use it. So the best option is to just define it as u64 to begin with.

I see no benefit to define it as u16 when we then convert it later anyways. It's not like it's saving us a measurable amount of memory.

File: alacritty.yml
Comment (kchibisov) [reply alacritty.yml:368]: I mean I'm not sure what to expect from interval 0, from what I can see it shouldn't blink at all, however it blink, since we theoretically can't provide an interval of zero, so I was thinking to clamping it to a value that we can handle, like 50ms or something. We can handle 4-5ms though I guess.

But we can keep it as 0 since I don't think it could mess up anything.

Comment (chrisduerr) [reply alacritty.yml:368]: I suppose it might be helpful to limit it just for debugging purposes. Since a rate of `0` ms might flood the event queue and cause problems. So I think I'll just limit it to something like `10ms` to make sure other than what is visible to the user, we don't run into any internal issues.

**Review by kchibisov — APPROVED** [review]

**Review by chrisduerr — APPROVED** [review]

**Pull request merged**
